src/server/snapshot.cc

```diff
@@ -253,24 +266,28 @@ bool SliceSnapshot::BucketSaveCb(PrimeIterator it) {
 }
 
 unsigned SliceSnapshot::SerializeBucket(DbIndex db_index, PrimeTable::bucket_iterator it) {
-  // Must be atomic because after after we call it.snapshot_version_ we're starting
+  // Must be atomic because after we call it.snapshot_version_ we're starting
   // to send incremental updates instead of serializing the whole bucket: We must not
   // send the update until the initial SerializeBucket is called.
   // Relying on the atomicity of SerializeBucket is Ok here because only one thread may handle this
   // bucket.
   FiberAtomicGuard fg;
   DCHECK_LT(it.GetVersion(), snapshot_version_);
 
+  bucket_ser_in_progress_ = true;
   // traverse physical bucket and write it into string file.
   serialize_bucket_running_ = true;
-  it.SetVersion(snapshot_version_);
   unsigned result = 0;
+  auto og = it;
 
   while (!it.is_done()) {
     ++result;
+    // might yield
     SerializeEntry(db_index, it->first, it->second, nullopt, serializer_.get());
     ++it;
   }
+  og.SetVersion(snapshot_version_);
```

Comment:
I'm not sure if this is interesting on its own, as:

> other fibers might check against the version and decide to move on instead of block and cause a bucket split

If that other fiber is older, it would have been called by now (we call by the order of snapshot versions), and if it's newer it wouldn't matter if the value is `snapshot_version_` or smaller (it can't be bigger). Furthermore, we shouldn't really get to the point in parallel in multiple fibers because we have the above (to-be-changed) polling.

Furthermore^2, if you set it to be after, we could actually serialize the same key twice in a rare case in which we call `OnDbChange`, block, but the `Traverse()` fiber continues and reaches this bucket. This will generate an invalid RDB, which Dragonfly will reject loading.